# Problems 6


Name: Denitsa Ilieva

In [ ]:
import numpy as np
import pandas as pd
import random

# Fill in any place that says 'Your code here'.
# You can implement auxiliary methods if you deem it proper. You may also change or extend functions, but:
# Make sure the code blocks at the end of each problem run as intended!!!

## Problem 1
Prepare data for multiclass classification from the iris dataset. It consists of 4-dimensional feature vectors for 50 data for each of three classes (0 = iris setosa, 1 = iris versicolor, 2 = iris virginica). Prepare a 120/30 split into train and test by holding out each fifth datapoint for testing.

In [ ]:
def read_iris(filename):
    """Reads the iris file into memory.

    Args:
        filename: string. The filename

    Returns:
        list of lists (one for each data point). Inner lists contain a numpy array (the features)
            and an integer (iris type).
    """
    # Your code here
    df = pd.read_csv(filename,header=None)

    df[4] = df[4].replace(['Iris-setosa'],0)
    df[4] = df[4].replace(['Iris-versicolor'],1)
    df[4] = df[4].replace(['Iris-virginica'],2)

    return df

def split_data(data):
    """Split the data into training and test sets.

    Args:
        data: Output format of read_iris()

    Returns:
        Tuple containing training set and test set (each in the input format)
    """
    # Your code here
    X = []
    y = []

    for index, rows in data.iterrows():
    # Create list for the current row
      if index%4==0:
        my_list =[rows[0], rows[1], rows[2], rows[3], rows[4]]
        y.append(my_list)
      else:
        my_list =[rows[0], rows[1], rows[2], rows[3], rows[4]]
        X.append(my_list)

    return X,y

In [ ]:
data = read_iris("iris.csv")

In [ ]:
data

0    1    2    3  4
0    5.1  3.5  1.4  0.2  0
1    4.9  3.0  1.4  0.2  0
2    4.7  3.2  1.3  0.2  0
3    4.6  3.1  1.5  0.2  0
4    5.0  3.6  1.4  0.2  0
..   ...  ...  ...  ... ..
145  6.7  3.0  5.2  2.3  2
146  6.3  2.5  5.0  1.9  2
147  6.5  3.0  5.2  2.0  2
148  6.2  3.4  5.4  2.3  2
149  5.9  3.0  5.1  1.8  2

[150 rows x 5 columns]

In [ ]:
train, test = split_data(data)

In [ ]:
train

[[4.9, 3.0, 1.4, 0.2, 0.0],
 [4.7, 3.2, 1.3, 0.2, 0.0],
 [4.6, 3.1, 1.5, 0.2, 0.0],
 [5.4, 3.9, 1.7, 0.4, 0.0],
 [4.6, 3.4, 1.4, 0.3, 0.0],
 [5.0, 3.4, 1.5, 0.2, 0.0],
 [4.9, 3.1, 1.5, 0.1, 0.0],
 [5.4, 3.7, 1.5, 0.2, 0.0],
 [4.8, 3.4, 1.6, 0.2, 0.0],
 [4.3, 3.0, 1.1, 0.1, 0.0],
 [5.8, 4.0, 1.2, 0.2, 0.0],
 [5.7, 4.4, 1.5, 0.4, 0.0],
 [5.1, 3.5, 1.4, 0.3, 0.0],
 [5.7, 3.8, 1.7, 0.3, 0.0],
 [5.1, 3.8, 1.5, 0.3, 0.0],
 [5.1, 3.7, 1.5, 0.4, 0.0],
 [4.6, 3.6, 1.0, 0.2, 0.0],
 [5.1, 3.3, 1.7, 0.5, 0.0],
 [5.0, 3.0, 1.6, 0.2, 0.0],
 [5.0, 3.4, 1.6, 0.4, 0.0],
 [5.2, 3.5, 1.5, 0.2, 0.0],
 [4.7, 3.2, 1.6, 0.2, 0.0],
 [4.8, 3.1, 1.6, 0.2, 0.0],
 [5.4, 3.4, 1.5, 0.4, 0.0],
 [5.5, 4.2, 1.4, 0.2, 0.0],
 [4.9, 3.1, 1.5, 0.2, 0.0],
 [5.0, 3.2, 1.2, 0.2, 0.0],
 [4.9, 3.6, 1.4, 0.1, 0.0],
 [4.4, 3.0, 1.3, 0.2, 0.0],
 [5.1, 3.4, 1.5, 0.2, 0.0],
 [4.5, 2.3, 1.3, 0.3, 0.0],
 [4.4, 3.2, 1.3, 0.2, 0.0],
 [5.0, 3.5, 1.6, 0.6, 0.0],
 [4.8, 3.0, 1.4, 0.3, 0.0],
 [5.1, 3.8, 1.6, 0.2, 0.0],
 [4.6, 3.2, 1.4, 0.2

In [ ]:
random.shuffle(train)

In [ ]:
train

[[5.8, 2.8, 5.1, 2.4, 2.0],
 [7.6, 3.0, 6.6, 2.1, 2.0],
 [6.0, 2.2, 4.0, 1.0, 1.0],
 [6.6, 3.0, 4.4, 1.4, 1.0],
 [6.9, 3.1, 5.4, 2.1, 2.0],
 [6.5, 3.2, 5.1, 2.0, 2.0],
 [5.5, 2.5, 4.0, 1.3, 1.0],
 [6.1, 3.0, 4.9, 1.8, 2.0],
 [6.4, 3.2, 4.5, 1.5, 1.0],
 [5.7, 2.8, 4.5, 1.3, 1.0],
 [5.7, 3.0, 4.2, 1.2, 1.0],
 [5.4, 3.7, 1.5, 0.2, 0.0],
 [5.7, 2.5, 5.0, 2.0, 2.0],
 [6.0, 3.4, 4.5, 1.6, 1.0],
 [7.7, 3.8, 6.7, 2.2, 2.0],
 [6.7, 3.1, 4.7, 1.5, 1.0],
 [5.4, 3.9, 1.7, 0.4, 0.0],
 [5.1, 3.7, 1.5, 0.4, 0.0],
 [5.1, 3.8, 1.6, 0.2, 0.0],
 [6.1, 2.9, 4.7, 1.4, 1.0],
 [5.1, 3.4, 1.5, 0.2, 0.0],
 [6.2, 2.8, 4.8, 1.8, 2.0],
 [5.6, 2.7, 4.2, 1.3, 1.0],
 [5.2, 2.7, 3.9, 1.4, 1.0],
 [6.7, 3.0, 5.0, 1.7, 1.0],
 [6.3, 2.9, 5.6, 1.8, 2.0],
 [6.1, 2.8, 4.7, 1.2, 1.0],
 [4.8, 3.0, 1.4, 0.3, 0.0],
 [4.9, 3.1, 1.5, 0.2, 0.0],
 [4.6, 3.2, 1.4, 0.2, 0.0],
 [6.3, 2.8, 5.1, 1.5, 2.0],
 [5.8, 2.7, 5.1, 1.9, 2.0],
 [6.1, 2.6, 5.6, 1.4, 2.0],
 [5.5, 2.4, 3.7, 1.0, 1.0],
 [7.9, 3.8, 6.4, 2.0, 2.0],
 [6.1, 3.0, 4.6, 1.4

In [ ]:
test

[[5.1, 3.5, 1.4, 0.2, 0.0],
 [5.0, 3.6, 1.4, 0.2, 0.0],
 [4.4, 2.9, 1.4, 0.2, 0.0],
 [4.8, 3.0, 1.4, 0.1, 0.0],
 [5.4, 3.9, 1.3, 0.4, 0.0],
 [5.4, 3.4, 1.7, 0.2, 0.0],
 [4.8, 3.4, 1.9, 0.2, 0.0],
 [5.2, 3.4, 1.4, 0.2, 0.0],
 [5.2, 4.1, 1.5, 0.1, 0.0],
 [5.5, 3.5, 1.3, 0.2, 0.0],
 [5.0, 3.5, 1.3, 0.3, 0.0],
 [5.1, 3.8, 1.9, 0.4, 0.0],
 [5.3, 3.7, 1.5, 0.2, 0.0],
 [6.9, 3.1, 4.9, 1.5, 1.0],
 [6.3, 3.3, 4.7, 1.6, 1.0],
 [5.0, 2.0, 3.5, 1.0, 1.0],
 [5.6, 2.9, 3.6, 1.3, 1.0],
 [6.2, 2.2, 4.5, 1.5, 1.0],
 [6.3, 2.5, 4.9, 1.5, 1.0],
 [6.8, 2.8, 4.8, 1.4, 1.0],
 [5.5, 2.4, 3.8, 1.1, 1.0],
 [5.4, 3.0, 4.5, 1.5, 1.0],
 [5.6, 3.0, 4.1, 1.3, 1.0],
 [5.8, 2.6, 4.0, 1.2, 1.0],
 [5.7, 2.9, 4.2, 1.3, 1.0],
 [6.3, 3.3, 6.0, 2.5, 2.0],
 [6.5, 3.0, 5.8, 2.2, 2.0],
 [6.7, 2.5, 5.8, 1.8, 2.0],
 [6.8, 3.0, 5.5, 2.1, 2.0],
 [6.5, 3.0, 5.5, 1.8, 2.0],
 [6.9, 3.2, 5.7, 2.3, 2.0],
 [6.7, 3.3, 5.7, 2.1, 2.0],
 [6.4, 2.8, 5.6, 2.1, 2.0],
 [6.4, 2.8, 5.6, 2.2, 2.0],
 [6.3, 3.4, 5.6, 2.4, 2.0],
 [6.7, 3.1, 5.6, 2.4

## Problem 2
Implement a perceptron algorithm for multiclass classification using a block feature vector construction and apply it to the iris dataset.

In [ ]:
class MCP:
    """Multiclass Perceptron"""

    def __init__(self, features=4, labels=3, epochs=250):
        self.f = features
        self.l = labels
        self.N = epochs
        self.weights = None

    def initialize_weights(self):
        """Set self.weights to a numpy array. May contain zeros, or random numbers.
        """
        # Your code here
        #self.weights = np.random.rand(4)
        self.weights = {c: np.array([random.random() for _ in range(self.f)]) for c in range(self.l)}
        #self.weights = {c: np.array([0 for _ in range(self.f)]) for c in range(self.l)}
        return self

    def train(self, train):
        """Train the MCP on training data.

        Args:
            train: Iris data format. The training data
        """
        # Your code here.
        for _ in range(self.N):
            for feature_vect, category in zip([i[:4] for i in train],[i[4] for i in train]):
                print('feature dict', feature_vect)
                category = category.astype(int)
                print(category)


                # Initialize arg_max value, predicted class.
                arg_max, predicted_class = 0, 0

                # Multi-Class Decision Rule:
                for c in range(self.l):
                    print('weights',self.weights)
                    current_activation = np.dot(feature_vect, self.weights[c])
                    print('current activation',current_activation)
                    if current_activation >= arg_max:
                        arg_max, predicted_class = current_activation, c

                # Update Rule:
                print('predicted class',predicted_class)
                if not (category == predicted_class):
                    self.weights[category] = self.weights[category]+ feature_vect
                    self.weights[predicted_class] = self.weights[predicted_class]- feature_vect
                print('final weights', self.weights)

    def predict(self, feature_dict):
        """
        Categorize a brand-new, unseen data point based on the existing collected data.
        :param  feature_dictionary  Dictionary of the same form as the training feature data.
        :return                     Return the predicted category for the data point.
        """


        # Initialize arg_max value, predicted class.
        arg_max, predicted_class = 0, 0

        # Multi-Class Decision Rule:
        for c in range(self.l):
            current_activation = np.dot(feature_dict, self.weights[c])
            if current_activation >= arg_max:
                arg_max, predicted_class = current_activation, c

        return predicted_class

    def test(self, test):
        """Test the MCP on test data.

        Args:
            test: Iris data format. The test data

        Returns:
            float -- Accuracy on the test set
        """
        # Your code here
        correct, incorrect = 0, 0
        for feature_vect, category in zip([i[:4] for i in test],[i[4] for i in test]):
            actual_class = category
            predicted_class = self.predict(feature_vect)

            if actual_class == predicted_class:
                correct += 1
            else:
                incorrect += 1

        print("ACCURACY:")
        print("Model Accuracy:", (correct * 1.0) / ((correct + incorrect) * 1.0))

## Problem 3
Evaluate your trained models on the training set and test set. What is the classification error on the test set? Can you ensure zero error on the training set? Explain your findings!

In [ ]:
mcp = MCP()
mcp.initialize_weights()

In [ ]:
mcp.train(train)

Die letzten 5000 Zeilen der Streamingausgabe wurden abgeschnitten.
feature dict [4.8, 3.4, 1.6, 0.2]
0
weights {0: array([  45.87780348,   94.39053221, -107.37618798,  -51.87300009]), 1: array([ 35.72143531,  56.21601916, -39.72944927, -94.62314099]), 2: array([ -80.83627224, -149.99479622,  148.61946364,  148.01319853])}
current activation 358.9647654242134
weights {0: array([  45.87780348,   94.39053221, -107.37618798,  -51.87300009]), 1: array([ 35.72143531,  56.21601916, -39.72944927, -94.62314099]), 2: array([ -80.83627224, -149.99479622,  148.61946364,  148.01319853])}
current activation 280.10560761135383
weights {0: array([  45.87780348,   94.39053221, -107.37618798,  -51.87300009]), 1: array([ 35.72143531,  56.21601916, -39.72944927, -94.62314099]), 2: array([ -80.83627224, -149.99479622,  148.61946364,  148.01319853])}
current activation -630.6026323684976
predicted class 0
final weights {0: array([  45.87780348,   94.39053221, -107.37618798,  -51.87300009]), 1: array([ 35.72

Results for epochs = 250, $w^{0}$ contains only zeros, no shuffle on the train data

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.7105263157894737


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.7053571428571429


With random.shuffle of the test data, epochs=250 and random weight values for $w^{0}$

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 1.0


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.9285714285714286


Results for 100 epochs, weights = [0,0,0,0]

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.39473684210526316


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.4107142857142857


With random.shuffle on the train data, epochs=100 and random weight values for $w^{0}$

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.9473684210526315


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.9732142857142857


Results for epochs = 50, weights = [0,0,0,0]

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.7631578947368421


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.8035714285714286


With random.shuffle of the train data, epochs=50 and random weight values for $w^{0}$

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.9210526315789473


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.9821428571428571


With random.shuffle of the test data and epochs=150

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.7105263157894737


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.7857142857142857


With random.shuffle of the test data, epochs=150 and random weight values for $w^{0}$

In [ ]:
mcp.test(test)

ACCURACY:
Model Accuracy: 0.9473684210526315


In [ ]:
mcp.test(train)

ACCURACY:
Model Accuracy: 0.9732142857142857


### Play around: Adjust learning rates, number of epochs, shuffle the training data, ...
### By shuffle the training data and initializing the $w^{0}$ with random numbers, I achieve much better results in comparison without shuffle the training data and initializing the $w^{0}$ with zeros. The evaluotion has shown that we can also get good results with 50 epochs (certainly if our data set is not that large). 250 epochs are far too many - we even come to overfitting in the test with shuffled training data and $w^{0}$ equal random numbers -> in this case we can add regulation